libcli.c

```diff
@@ -1362,49 +1301,134 @@ int cli_loop(struct cli_def *cli, int sockfd) {
 
       /* TAB completion */
       if (c == CTRL('I')) {
-        char *completions[CLI_MAX_LINE_WORDS];
-        int num_completions = 0;
+        struct cli_comphelp comphelp = {0};
 
         if (cli->state == STATE_LOGIN || cli->state == STATE_PASSWORD || cli->state == STATE_ENABLE_PASSWORD) continue;
 
         if (cursor != l) continue;
 
-        num_completions = cli_get_completions(cli, cmd, completions, CLI_MAX_LINE_WORDS);
-        if (num_completions == 0) {
+        cli_get_completions(cli, cmd, c, &comphelp);
+        if (comphelp.num_entries == 0) {
           _write(sockfd, "\a", 1);
-        } else if (num_completions == 1) {
-          // Single completion
-          for (; l > 0; l--, cursor--) {
-            if (cmd[l - 1] == ' ' || cmd[l - 1] == '|') break;
-            _write(sockfd, "\b", 1);
-          }
-          strcpy((cmd + l), completions[0]);
-          l += strlen(completions[0]);
-          cmd[l++] = ' ';
-          cursor = l;
-          _write(sockfd, completions[0], strlen(completions[0]));
-          _write(sockfd, " ", 1);
         } else if (lastchar == CTRL('I')) {
           // double tab
           int i;
-          _write(sockfd, "\r\n", 2);
-          for (i = 0; i < num_completions; i++) {
-            _write(sockfd, completions[i], strlen(completions[i]));
-            if (i % 4 == 3)
+          for (i = 0; i < comphelp.num_entries; i++) {
+            if (i % 4 == 0)
               _write(sockfd, "\r\n", 2);
             else
-              _write(sockfd, "     ", 1);
+              _write(sockfd, " ", 1);
+            _write(sockfd, comphelp.entries[i], strlen(comphelp.entries[i]));
           }
-          if (i % 4 != 3) _write(sockfd, "\r\n", 2);
+          _write(sockfd, "\r\n", 2);
           cli->showprompt = 1;
-        } else {
-          // More than one completion
+        } else if (comphelp.num_entries == 1) {
+          // Single completion - show *unless* the optional/required 'prefix' is present
+          if ((comphelp.entries[0][0] != '[') && (comphelp.entries[0][0] != '<')) {
+            for (; l > 0; l--, cursor--) {
+              if (cmd[l - 1] == ' ' || cmd[l - 1] == '|' || (comphelp.comma_separated && cmd[l - 1] == ',')) break;
+              _write(sockfd, "\b", 1);
+            }
+            strcpy((cmd + l), comphelp.entries[0]);
+            l += strlen(comphelp.entries[0]);
+            cmd[l++] = ' ';
+            cursor = l;
+            _write(sockfd, comphelp.entries[0], strlen(comphelp.entries[0]));
+            _write(sockfd, " ", 1);
+            // and now forget the tab, since we just found a single match
+            lastchar = '\0';
+          } else {
+            // Yes, we had a match, but it wasn't required - remember the tab in case the user double tabs....
+            lastchar = CTRL('I');
+          }
+        } else if (comphelp.num_entries > 1) {
+          /* More than one completion
```

Comment:
Here's where my obligatory note on comment consistency goes. There's a lack of consistency, some comments have full sentences, some have lowercase first characters. Some are multi-line with the first line blank and some aren't.

I'd like to see:
* Single-line comments should be as close as possible to sentences - Uppercase first character, full-stop at the end if it makes sense.
* Multi-line comments should have a blank first line:
```
/*
  * This is the comment.
  */
```

I am very aware that the existing code is _the worst_ in regards to this. Again, "go away Dave" is fine, I'll fix it another day.

Reply:
Concur - and I'm notoriously inconsistent myself.  Had coworker whose comment skills were almost perfectly consistent, accurent, and pertinent.  I am not that guy....